packages/protocol/test/layout/layout.ts

```diff
@@ -0,0 +1,297 @@
+    it('reports a struct change', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original,
+        testCases.typechange_in_library_struct
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/struct.*changed/])
+    })
+  })
+
+  describe('when a field is removed from a struct', () => {
+    it('reports a struct change', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original,
+        testCases.removed_from_struct
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/struct.*changed/])
+    })
+  })
+
+  describe('when a field is removed from a library struct', () => {
+    it('reports a struct change', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original,
+        testCases.removed_from_library_struct
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/struct.*changed/])
+    })
+  })
+
+  describe('when a field is inserted in a library struct', () => {
+    it('reports a struct change', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original,
+        testCases.inserted_in_library_struct
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/struct.*changed/])
+    })
+  })
+
+  describe('when a fixed array has length increased', () => {
+    it('reports a typechanged variable', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original_complex,
+        testCases.longer_fixed_array
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/had type/])
+    })
+  })
+
+  describe('when a fixed array has length decreased', () => {
+    it('reports a typechanged variable', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original_complex,
+        testCases.shorter_fixed_array
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/had type/])
+    })
+  })
+
+  describe('when a fixed array becomes dynamic', () => {
+    it('reports a typechanged variable', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original_complex,
+        testCases.fixed_to_dynamic_array
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/had type/])
+    })
+  })
+
+  describe('when a dynamic array becomes fixed', () => {
+    it('reports a typechanged variable', () => {
+      const report = reportLayoutIncompatibilities(
+        testCases.original_complex,
+        testCases.dynamic_to_fixed_array
+      )
+      assertNotCompatible(report)
+      assertContractErrorsMatch(report, 'TestContract', [/had type/])
+    })
+  })
+
+  // TODO(m-chrzan): @openzeppelin/upgrades erases information about mapping key
```

Comment:
I was very confused until I saw `describe.skip` 😄 

+1 to patching this behavior, as changing the key type should be considered backwards incompatible. I think it's fine to rely on PR reviews and audits to catch these changes for now.